File: server/graph_service/auth_middleware.py
"""
API Authentication Middleware for GraphiTi FastAPI Server

Implements Bearer Token and API Key authentication for public deployment.
"""

from fastapi import Request, HTTPException, status
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
import logging

from graph_service.config import Settings

logger = logging.getLogger(__name__)


class AuthMiddleware(BaseHTTPMiddleware):
    """Authentication middleware for FastAPI."""

    def __init__(self, app, settings: Settings):
        super().__init__(app)
        self.settings = settings
        self.public_endpoints = [
            endpoint.strip()
            for endpoint in settings.api_auth_public_endpoints.split(',')
        ]

    async def dispatch(self, request: Request, call_next):
        """Process request and validate authentication."""

        # Skip authentication if disabled
        if not self.settings.api_auth_enabled:
            return await call_next(request)

        # Check if this is a public endpoint
        if self._is_public_endpoint(request.url.path):
            logger.debug(f"Public endpoint accessed: {request.url.path}")
            return await call_next(request)

        # Validate authentication
        try:
            if self.settings.api_auth_method == 'bearer':
                self._validate_bearer_token(request)
            elif self.settings.api_auth_method == 'apikey':
                self._validate_api_key(request)
            else:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail=f"Invalid auth method: {self.settings.api_auth_method}"
                )

            # Authentication successful
            return await call_next(request)

        except HTTPException as e:
            logger.warning(
                f"Authentication failed for {request.method} {request.url.path}: {e.detail}"
            )

            # Return authentication error
            if self.settings.api_auth_method == 'bearer':
                return JSONResponse(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    content={"detail": e.detail},
                    headers={"WWW-Authenticate": "Bearer realm=\"GraphiTi API\""}
                )
            else:
                return JSONResponse(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    content={"detail": e.detail}
                )

    def _is_public_endpoint(self, path: str) -> bool:
        """Check if the path is a public endpoint."""
        return any(
            path == endpoint or path.startswith(f"{endpoint}/")
            for endpoint in self.public_endpoints
        )

    def _validate_bearer_token(self, request: Request) -> None:
        """Validate Bearer token authentication."""
        auth_header = request.headers.get("Authorization")

        if not auth_header:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Missing Authorization header"
            )

        if not auth_header.startswith("Bearer "):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid Authorization header format. Expected: Bearer <token>"
            )

        token = auth_header[7:].strip()  # Remove "Bearer " prefix

        if not token:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Empty bearer token"
            )

        if token != self.settings.api_auth_bearer_token:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid bearer token"
            )

        logger.debug("Bearer token validated successfully")

    def _validate_api_key(self, request: Request) -> None:
        """Validate API Key authentication."""
        api_key = request.headers.get("X-API-Key")

        if not api_key:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Missing X-API-Key header"
            )

        if not api_key.strip():
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Empty API key"
            )

        if api_key != self.settings.api_auth_api_key:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid API key"
            )

        logger.debug("API key validated successfully")

File: mcp-http-server/src/auth-middleware.ts
/**
 * Authentication Middleware for MCP Server
 * Implements MCP Standard OAuth 2.1 Bearer Token Authentication
 *
 * Supports:
 * - Bearer Token authentication (recommended)
 * - API Key authentication (simpler alternative)
 * - Public endpoints (no auth required)
 */

import type { Request, Response, NextFunction } from 'express';
import type { GraphitiConfig, Logger } from './config.js';

/**
 * Create authentication middleware based on configuration
 */
export function createAuthMiddleware(config: GraphitiConfig, logger: Logger) {
  return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    // Skip authentication if disabled
    if (!config.auth.enabled) {
      return next();
    }

    // Check if this is a public endpoint
    const isPublicEndpoint = config.auth.publicEndpoints.some(
      (endpoint) => req.path === endpoint || req.path.startsWith(`${endpoint}/`)
    );

    if (isPublicEndpoint) {
      logger.debug(`Public endpoint accessed: ${req.path}`);
      return next();
    }

    // Validate authentication based on configured method
    try {
      if (config.auth.method === 'bearer') {
        await validateBearerToken(req, config, logger);
      } else if (config.auth.method === 'apikey') {
        await validateApiKey(req, config, logger);
      }

      // Authentication successful
      next();
    } catch (error) {
      // Authentication failed
      const errorMessage = error instanceof Error ? error.message : 'Authentication failed';
      logger.warn(`Authentication failed for ${req.method} ${req.path}: ${errorMessage}`);

      // Return MCP-compliant error response
      if (config.auth.method === 'bearer') {
        res.status(401).json({
          error: 'unauthorized',
          message: errorMessage,
        });
        res.setHeader('WWW-Authenticate', 'Bearer realm="MCP Server"');
      } else {
        res.status(401).json({
          error: 'unauthorized',
          message: errorMessage,
        });
      }
    }
  };
}

/**
 * Validate Bearer Token (OAuth 2.1 standard)
 *
 * Expected header format: Authorization: Bearer <token>
 */
async function validateBearerToken(
  req: Request,
  config: GraphitiConfig,
  logger: Logger
): Promise<void> {
  const authHeader = req.headers.authorization;

  if (!authHeader) {
    throw new Error('Missing Authorization header');
  }

  if (!authHeader.startsWith('Bearer ')) {
    throw new Error('Invalid Authorization header format. Expected: Bearer <token>');
  }

  const token = authHeader.substring(7).trim(); // Remove "Bearer " prefix

  if (!token) {
    throw new Error('Empty bearer token');
  }

  // Validate token against configured value
  if (token !== config.auth.bearerToken) {
    throw new Error('Invalid bearer token');
  }

  logger.debug('Bearer token validated successfully');
}

/**
 * Validate API Key
 *
 * Expected header format: X-API-Key: <key>
 */
async function validateApiKey(
  req: Request,
  config: GraphitiConfig,
  logger: Logger
): Promise<void> {
  const apiKey = req.headers['x-api-key'] as string | undefined;

  if (!apiKey) {
    throw new Error('Missing X-API-Key header');
  }

  if (!apiKey.trim()) {
    throw new Error('Empty API key');
  }

  // Validate API key against configured value
  if (apiKey !== config.auth.apiKey) {
    throw new Error('Invalid API key');
  }

  logger.debug('API key validated successfully');
}

/**
 * Utility function to check if request is authenticated
 * Can be used in route handlers for additional checks
 */
export function isAuthenticated(req: Request, config: GraphitiConfig): boolean {
  if (!config.auth.enabled) {
    return true; // Auth disabled, consider all requests authenticated
  }

  try {
    if (config.auth.method === 'bearer') {
      const authHeader = req.headers.authorization;
      if (!authHeader?.startsWith('Bearer ')) return false;
      const token = authHeader.substring(7).trim();
      return token === config.auth.bearerToken;
    } else if (config.auth.method === 'apikey') {
      const apiKey = req.headers['x-api-key'] as string;
      return apiKey === config.auth.apiKey;
    }
  } catch {
    return false;
  }

  return false;
}
